File: app/src/main/java/com/ternaryop/utils/JSONUtils.java
package com.ternaryop.utils;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import org.json.JSONTokener;

public class JSONUtils {
    public static Object toJSON(Object object) throws JSONException {
        if (object instanceof Map) {
            JSONObject json = new JSONObject();
            Map<?, ?> map = (Map<?, ?>) object;
            for (Object key : map.keySet()) {
                json.put(key.toString(), toJSON(map.get(key)));
            }
            return json;
        } else if (object instanceof Iterable) {
            JSONArray json = new JSONArray();
            for (Object value : ((Iterable<?>) object)) {
                json.put(value);
            }
            return json;
        } else {
            return object;
        }
    }

    public static boolean isEmptyObject(JSONObject object) {
        return object.names() == null;
    }

    public static Map<String, Object> getMap(JSONObject object, String key)
            throws JSONException {
        return toMap(object.getJSONObject(key));
    }

    @SuppressWarnings("unchecked")
    public static Map<String, Object> toMap(JSONObject object)
            throws JSONException {
        Map<String, Object> map = new HashMap<String, Object>();
        Iterator<String> keys = object.keys();
        while (keys.hasNext()) {
            String key = keys.next();
            map.put(key, fromJson(object.get(key)));
        }
        return map;
    }

    public static List<Object> toList(JSONArray array) throws JSONException {
        List<Object> list = new ArrayList<Object>();
        for (int i = 0; i < array.length(); i++) {
            list.add(fromJson(array.get(i)));
        }
        return list;
    }

    private static Object fromJson(Object json) throws JSONException {
        if (json == JSONObject.NULL) {
            return null;
        } else if (json instanceof JSONObject) {
            return toMap((JSONObject) json);
        } else if (json instanceof JSONArray) {
            return toList((JSONArray) json);
        } else {
            return json;
        }
    }

    public static JSONObject jsonFromInputStream(InputStream is) throws JSONException, IOException {
        BufferedReader in = new BufferedReader(new InputStreamReader(is));
        StringBuilder sb = new StringBuilder();
        String line;
        while ((line = in.readLine()) != null) {
            sb.append(line);
        }
        return (JSONObject) new JSONTokener(sb.toString()).nextValue();
    }
}

File: library/src/main/java/com/ternaryop/utils/json/JSONUtils.kt
package com.ternaryop.utils.json

import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.util.ArrayList
import java.util.HashMap

val JSONObject.isEmptyObject: Boolean
    get() = names() == null

@Throws(JSONException::class)
@Suppress("ReturnCount")
fun Any.toJSON(): Any {
    when {
        this is Map<*, *> -> {
            val json = JSONObject()
            for (key in keys) {
                json.put(key.toString(), this.get(key)!!.toJSON())
            }
            return json
        }
        this is Iterable<*> -> {
            val json = JSONArray()
            for (value in this) {
                json.put(value)
            }
            return json
        }
        else -> return this
    }
}

@Throws(JSONException::class)
fun JSONObject.toMap(): Map<String, Any> {
    val map = HashMap<String, Any>()
    val keys = keys()
    while (keys.hasNext()) {
        val key = keys.next()
        get(key).fromJson()?.let { map[key] = it }
    }
    return map
}

@Throws(JSONException::class)
fun Any.fromJson(): Any? {
    return when {
        this === JSONObject.NULL -> null
        this is JSONObject -> toMap()
        this is JSONArray -> toList()
        else -> this
    }
}

@Throws(JSONException::class)
fun JSONArray.toList(): List<Any> {
    val list = ArrayList<Any>()
    for (i in 0 until length()) {
        get(i).fromJson()?.let { list.add(it) }
    }
    return list
}

@Throws(JSONException::class)
fun JSONObject.getMap(key: String): Map<String, Any> = getJSONObject(key).toMap()

